### ISSUE TITLE
Python 3.11 compatibility

### ISSUE THREAD

**Issue by pekkaklarck** [robotframework/robotframework#4401]
Python 3.11 is getting ready and we need to test that RF 5.1 is compatible with it. That requires:

- [x] Running all tests with the latest Python 3.11 version (beta 4 at the moment) both on Linux.
- [x] Same as above on Windows.
- [x] Fixing possible issues.
- [x] Adding Python 3.11 support to project metadata in setup.py.
- [x] Enabling Python 3.11 runs on CI (assuming it is already supported there).

**Comment by pekkaklarck** [comment]
Some tests failed. I analyzed why they failed and these were the reasons:

1. Python 3.11 has enhanced error reporting and as a part of that tracebacks have `^^^^` markers showing where errors occur in more detail. This breaks some unit tests, but they are easy to fix by filtering such lines away from tracebacks before comparing them.
2. `typing.get_type_hints` doesn't anymore automatically add `None` to the returned types if an argument has `None` as a default value. This means that if you have a keyword with a signature like `arg: str = None`, `arg` only has `str` as a type nowadays when earlier it had both `str` and `None` (i.e. `NoneType`). This means that if you call the keyword with a `None` object, it is not anymore accepted as-is but instead converted to a string `None´. I don't like out conversion behavior depending on Python versions, so I'll adapt the code so that `None` isn't converted if an argument has `None` as a default. 
3. Python 3.11 is more strict about embedding flags to regexp patterns. With earlier versions, for example, `^(?s)x` works but it causes and error with Python 3.11. Luckily the only place in the code this error occurs is`String.Get Lines Matching Regexp` keyword and fixing it is easy.

**Comment by pekkaklarck** [comment]
This is not done.

Getting tests to pass on CI turned out to be somewhat hard. Strangely more tests validating tracebacks failed on CI due to those `^^^` lines added by Python 3.11. CI uses beta 3 while I tested locally with beta 4, but it would be a strange if that affected this behavior. Anyway, I got tests passing also on CI by removing `^^^` lines.

Also needed to disable testing using Python 3.11 on Windows. Too many dependencies lacked Python 3.11 support.
